orangecontrib/spectroscopy/preprocess/__init__.py

```diff
@@ -721,3 +721,69 @@ def __call__(self, data):
         domain = Orange.data.Domain(atts, data.domain.class_vars,
                                     data.domain.metas)
         return data.from_table(domain, data)
+
+class DespikeFeature(SelectColumn):
+    pass
+
+class _DespikeCommon(CommonDomainOrderUnknowns):
+    def __init__(self, threshold, cutoff, dis, domain):
+        super().__init__(domain)
+        self.threshold = threshold
+        #threshold sets up a limit for the modified_z_scores test cutoff for spikes
+        self.cutoff = cutoff
+        #cutoff allows for spiked spectra to be determined from a large
+        #multispectral image
+        self.dis = dis
+        #dis sets the distance over which to interpolate spiked areas
+
+    def transformed(self, data, X):
+        def interpolatespikes(spikes, row):
+            for i in np.arange(len(spikes)):
+                if spikes[i] != 0:
+                    #need if statements to ensure no conflicts with edge points
+                    if (i+self.dis) >= len(spikes):
+                        w = np.arange(i-self.dis)
+                    else:
+                        w = np.arange(i - self.dis, i + self.dis)
+                    w2 = w[spikes[w] == 0]
+                    row_out[i] = np.mean(row[w2])
+            return out.append(row_out)
+        out = []
+        Series = np.array(data)
+        if Series.size > 0:
+            for row in Series:
+                Distance = np.diff(row, n=0)
+                #Spiked spectra are processed and non spiked are passed through
+                if np.any(Distance > self.cutoff):
+                    median1 = np.median(row)
+                    mad_int = np.median([np.abs(row - median1)])
+                    modified_z_scores = 0.6745 * (row - median1) / mad_int
+                    difference = (abs(np.array(modified_z_scores)) > self.threshold)
+                    row_out = row.copy()
```

Comment:
row_out should be a parameter of interpolatespikes. To modify an external (semi-global) variable within a function is discouraged. Also, interpolatespikes should not modify the array out per the same principle.